params/src/dictionary.cpp

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (C) 1998-2017 ALPS Collaboration. See COPYRIGHT.TXT
+ * All rights reserved. Use is subject to license terms. See LICENSE.TXT
+ * For use in publications, see ACKNOWLEDGE.TXT
+ */
+
+/** @file dictionary.cpp
+    Contains implementation of some alps::params_ns::dictionary members */
+
+#include <alps/dictionary.hpp>
+
+#include <alps/hdf5/map.hpp>
+
+#ifdef ALPS_HAVE_MPI
+#include <alps/utilities/mpi_map.hpp>
+#endif
+
+namespace alps {
+    namespace params_ns {
+
+        dictionary::map_type::const_iterator dictionary::find_nonempty_(const std::string& key) const {
+            map_type::const_iterator it=map_.find(key);
+            if (it!=map_.end() && !(it->second).empty())
+                return it;
+            else
+                return map_.end();
+        }
+
+        
+        /// Access with intent to assign
+        dictionary::value_type& dictionary::operator[](const std::string& key) {
+            map_type::iterator it=map_.lower_bound(key);
+            if (it==map_.end() ||
+                map_.key_comp()(key, it->first)) {
+                // it's a new element, we have to construct it here
+                // and copy it to the map, returning the ref to the inserted element
+                it=map_.insert(it,map_type::value_type(key, value_type(key)));
+            }
+            // return reference to the existing or the newly-created element
+            return it->second;
+        }
+
+        /// Read-only access
+        const dictionary::value_type& dictionary::operator[](const std::string& key) const {
```

Comment:
I understand this. It may throw on accessing the "empty value" that was silently inserted here, or, in this example, it will likely output the string representation of the empty value.

On the other hand, if the user will deliberately works with constant parameter objects, the user will get an exception of "unknown key" on access. My `operator[]` is analogous to C++11 `at()`.

Reply:
> On the other hand, if the user will deliberately works with constant parameter objects, the user will get an exception of "unknown key" on access. My operator[] is analogous to C++11 at().

I think this is a question of convenience.
What should I do with my example to ensure that `operator[]` throws for unknown keys?
I can see two options.
```c++
alps::params p;
p.define<int>("my_param", 42, "My parameter");
std::cout << const_cast<alps::params const&>(p)["invalid_param"] << std::endl;
```
```c++
const alps::params p;
const_cast<alps::params &>(p).define<int>("my_param", 42, "My parameter");
std::cout << p["invalid_param"] << std::endl;
```
Both are ugly, especially if you have multiple calls to `define()` and/or `operator[]`.

Reply:
Eh, what about this?

```c++
void define_my_params(alps::params& p) {
  p.define<int>("my_param", 42, "My parameter")
   .define<double>("my_other_param", 42.25, "My other parameter")
   .define<std::string>("my_third_param", "whatever", "My third parameter")
  ;
}
void use_my_params(const alps::dictionary& d) {
    std::cout << "My int is " << d["my_param"];
    //.... and so on ....
    std::cout << "This won't work: " << d["unknown_param"];
}

int main(int argc, char**argv) {
    alps::params p(argc,argv);
    define_my_params(p);
    use_my_params(p);
}
```

I'd like to know a good solution to this problem. 
There is no way to distinguish whether `operator[]` was called at the left side or right side of an assignment:
```c++
T x=p[key];
p[key]=x;
```
The only alternative i am aware of is to have `p.operator[](key)` construct a proxy object containing references to both `p` and `key`; this object would know whether to call `p.get(key)` (in `proxy_t::operator T()`) or `p.assign(key,x)` (in `proxy_t::operator=(const T&)`). Unfortunately, this also has problems: the proxy object may outlive either `p` or `key`. Granted, `key` may be copied; `p` may actually be a ref-counted shared pointer to implementation object in the heap (do i make sense?). Is following this route worth the outcome?

Reply:
Using an auxiliary function is, of course, possible but less convenient.
I actually like the idea to add a new method `assign<T>(key, x)` and to make all overloads of `operator[](key)` throw.

Reply:
The scoping problem is not exclusive to the proxy object: you have exactly the same problem with the usual `operator[]` already returns a reference that may outlive its host.

AFAI understand you correctly, you are proposing something like a `Maybe` monad? That solution I like.  I always considered the automatic default construction of non-existent values in `std::map` a bug.
```C++
template <typename Container>
class maybe_value {
public:
    typedef typename Container::key_type key_type;
    typedef typename Container::mapped_type mapped_type;

    maybe_value(Container &, const key_type &);
    operator mapped_type &();      // throws if non-existent key
private:
    Container &cont_;
    key_type key_;
};
```